### PR TITLE
Allow dApps to suggest change of RPC (EIP-3085 EIP-3326)

### LINKED ISSUE #11856: Allow dApps to suggest change of RPC (EIP-3085)
As a user, I want a dApp to be able to propose to me to switch the ethereum chain my wallet (and thus browser experience) is interacting with while I am browsing the dapps. 

Metamask will use this EIP to do this https://eips.ethereum.org/EIPS/eip-3085 and we should follow the same standard because it's de facto the one that dapps will use

Considerations:
- security first. This is a sensitive feature and security considerations should be analysed in detail. See EIP-3085 for details https://eips.ethereum.org/EIPS/eip-3085

We should mimic metamask behavior when it comes to alerts:

confirmation-0

confirmation-1

### Notes

Changing network without login/logout is currently not supported, but it's still worth working on this feature as it will eventually land in status-go.
For now, we should notify the user that it will require login/logout, and trigger a switch of network and login/logoutonce the network has been added.

### PR DESCRIPTION
**Pull request by qfrank**
fixes #11856

notes: the modification for UI needs improvements, but this PR won't include the improvements(i've discussed with @cammellos before), because i'm not very familiar with UI implementation.

screenshots:
image
image

### CODE REVIEW

**Comment by flexsurfer** [comment]
@John-44 could we add this in new designs ?

**Comment by cammellos** [comment]
We can leave this under a flag if designs are not ready, but in the meantime it's good to test it, @qfrank could you please check `make lint`, I think linting is failing, thanks!

**Comment by qfrank** [comment]
> We can leave this under a flag if designs are not ready, but in the meantime it's good to test it, @qfrank could you please check `make lint`, I think linting is failing, thanks!

done :)

File: src/status_im/ui/screens/browser/permissions/views.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/browser/permissions/views.cljs]: i18n

File: src/status_im/ui/screens/browser/permissions/views.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/browser/permissions/views.cljs]: i18n

File: src/status_im/ui/screens/browser/permissions/views.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/browser/permissions/views.cljs]: i18n

File: src/status_im/ui/screens/browser/permissions/views.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/browser/permissions/views.cljs]:
```clojure
(-> description-template
       (string/replace "${{network-from}}" (:network-from params))
       (string/replace "${{network-to}}" (:network-to params)))
```

Comment (flexsurfer) [reply src/status_im/ui/screens/browser/permissions/views.cljs]: also :network-from and :network-to could be destructured

File: src/status_im/ui/screens/browser/permissions/views.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/browser/permissions/views.cljs]: probably its better to have a separate component

File: src/status_im/browser/permissions.cljs
Comment (flexsurfer) [inline src/status_im/browser/permissions.cljs]: just call event handler here instead of dispatch

File: src/status_im/browser/permissions.cljs
Comment (flexsurfer) [inline src/status_im/browser/permissions.cljs]: not used anymore ?

File: src/status_im/browser/core.cljs
Comment (flexsurfer) [inline src/status_im/browser/core.cljs]: probably better to move all chain related methods to separate ns

File: src/status_im/browser/permissions.cljs
Comment (flexsurfer) [inline src/status_im/browser/permissions.cljs]: why this condition inside send-response ? i believe permissioned-switch-ethereum-chain don't send any response ?

File: src/status_im/browser/permissions.cljs
Comment (flexsurfer) [inline src/status_im/browser/permissions.cljs]: why not in the supported-permissions ? and implemented separately in cond ?

**Comment by flexsurfer** [comment]
implementation little bit confusing, because we intercept request in clojure, and make it as an permission request, and intercept response , probably there is a better way to do it, we should try to find

File: src/status_im/ui/screens/browser/permissions/views.cljs
Comment (qfrank) [reply src/status_im/ui/screens/browser/permissions/views.cljs]: i thought it was relate to UI implementation, so i just simply put "Network Name" here ..

Comment (flexsurfer) [reply src/status_im/ui/screens/browser/permissions/views.cljs]: could you elaborate UI implementation, you mean it will change, but we can't merge it then?

Comment (flexsurfer) [reply src/status_im/ui/screens/browser/permissions/views.cljs]: i guess we still will have these fields visible to users? so still better to move them to i18n

File: src/status_im/browser/permissions.cljs
Comment (qfrank) [reply src/status_im/browser/permissions.cljs]: - because it's a `dynamic` [permission](https://github.com/qfrank/status-react/blob/d4ecbd6183ed67192e56e9953d8efca5b197cb44/src/status_im/browser/core.cljs#L374), the `chainId` is a variable 
- because i want to reuse [permission panel](https://github.com/qfrank/status-react/blob/d4ecbd6183ed67192e56e9953d8efca5b197cb44/src/status_im/ui/screens/browser/permissions/views.cljs#L45)

File: src/status_im/browser/permissions.cljs
Comment (qfrank) [reply src/status_im/browser/permissions.cljs]: because i want to reuse [permission panel](https://github.com/qfrank/status-react/blob/d4ecbd6183ed67192e56e9953d8efca5b197cb44/src/status_im/ui/screens/browser/permissions/views.cljs#L45)

File: src/status_im/browser/permissions.cljs
Comment (qfrank) [reply src/status_im/browser/permissions.cljs]: replaced with `send-response`, it's an enhancement so that we can get extra params after user allowed or denied

**Comment by qfrank** [comment]
Hi @flexsurfer , ready to review :)

File: src/status_im/browser/multi_chain.cljs
Comment (flexsurfer) [inline src/status_im/browser/multi_chain.cljs]: im not sure this still need this, we could move it to ui and drop type , just use `method` param

File: src/status_im/browser/multi_chain.cljs
Comment (flexsurfer) [inline src/status_im/browser/multi_chain.cljs]: you shouldn't use a subscription inside the event, you must get all data from event's db

File: src/status_im/browser/multi_chain.cljs
Comment (flexsurfer) [inline src/status_im/browser/multi_chain.cljs]: `:dispatch` effect should be used instead

File: src/status_im/browser/core.cljs
Comment (flexsurfer) [inline src/status_im/browser/core.cljs]: better to rename multi-chain to EIP-3085 or EIP-3326

File: src/status_im/constants.cljs
Comment (flexsurfer) [inline src/status_im/constants.cljs]: i think that we don't need this anymore, method could be used

**Comment by flexsurfer** [comment]
thank you @qfrank left a few comments

File: src/status_im/browser/eip3326.cljs
Comment (flexsurfer) [inline src/status_im/browser/eip3326.cljs]: could you elaborate pls why dispatch-and-chill is needed here? it can't be used here, because map should be result of this function, instead of dispatching `:eip3326.ui/dapp-permission-allowed` you have to use `dispatch-and-chill`

**Review by flexsurfer — COMMENTED** [review]
sorry for the delay, a lost a track of this issue, lets move it to testing

**Comment by pavloburykh** [comment]
Hi, @qfrank! I will move PR to the Review column as currently it is not approved and thus is not ready for testing yet. As soon as it will be approved and all review comments are resolved, please move it back to E2E column. Thank you!

**Review by flexsurfer — APPROVED** [review]

**Comment by qoqobolo** [comment]
Hello @qfrank, thank you for PR!

Tested:
- Avalanche with https://traderjoexyz.com/home#/ dapp
- Matic (Polygon) with https://app.gamee.com/ dapp
- BSC Network with https://pancakeswap.finance/ dapp
- https://app.uniswap.org/#/swap for switching back to Mainnet
- denial of the request at each of the steps
- on Android and iOS devices
- regular and keycard accounts

Looks good to me except for one thing: switching to xDai (Gnosis) network does not work correctly using this functionality. 

**Steps:**
- open some Gnosis dapp in browser (i.e. https://honeyswap.1hive.eth.limo//#/swap or https://thehashavatars.com)
- connect wallet

**Expected result:** user is prompted to switch network
**Actual result:** nothing happens

If I switch the network to xDai manually I see my wallet connected in the DApp.


https://user-images.githubusercontent.com/67952253/195576684-aa8690fc-3e24-4300-93b0-9f2d787d8023.MP4


Maybe that's because the RPC URL has been updated again: https://docs.gnosischain.com/tools/wallets/metamask.

You still can switch to this network manually from the settings despite the fact that we still have https://gnosischain-rpc.gateway.pokt.network URL while the new one is https://rpc.gnosischain.com

Can you check please @qfrank?
@flexsurfer should we also update this URL in our networks list? I got an error 'Network request failed' a couple of times when switching manually, but couldn't reproduce

**Comment by flexsurfer** [comment]
> @flexsurfer should we also update this URL in our networks list? I got an error 'Network request failed' a couple of times when switching manually, but couldn't reproduce

its hard to say, probably we could create an issue, so we could investigate more

**Comment by qfrank** [comment]
> 

Hi @qoqobolo , i've checked https://honeyswap.1hive.eth.limo//#/swap and https://thehashavatars.com/ , looks like they won't call wallet_addEthereumChain/wallet_switchEthereumChain neither, so user need switch network manually, in another word, these dapps don't use EIP-3085/EIP-3326

**Comment by qoqobolo** [comment]
@qfrank got it, thanks for checking and for your work!
PR can be merged.

**Pull request merged**
